java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -35,41 +37,72 @@ import kotlinx.coroutines.withContext
  */
 class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
     val storageKey: StorageKey,
-    val backingType: Type,
-    val callbackFactory: (String) -> ProxyCallback<Data, Op, T>
+    val backingType: Type
 ) {
     private val storeMutex = Mutex()
     private val log = TaggedLog { "DirectStoreMuxer" }
+    private val proxyManager = RandomProxyCallbackManager<Data, Op, T>(
+        "backing",
+        Random
+    )
 
     // TODO(b/158262634): Make this CacheMap Weak.
     /* internal */ val stores = LruCacheMap<String, StoreRecord<Data, Op, T>>(
         50,
         livenessPredicate = { _, sr -> !sr.store.closed }
-    ) { _, sr -> closeStore(sr) }
+    ) { muxId, sr -> closeStore(muxId, sr) }
+
+    // Keeps track of all the stores a callbackId is registered to. Used when removing an
+    // observer (off method), as that requires unregistering to each store it is registered with.
+    private val callbackIdToMuxIdMap = mutableMapOf<Int, MutableSet<String>>()
+
+    fun on(callback: ProxyCallback<Data, Op, T>, callbackToken: Int? = null): Int {
+        synchronized(proxyManager) {
+            val callbackId = proxyManager.register(callback, callbackToken)
+            callbackIdToMuxIdMap[callbackId] = mutableSetOf()
+            return callbackId
+        }
+    }
+
+    fun off(callbackToken: Int) {
+        for (muxId in callbackIdToMuxIdMap[callbackToken]!!) {
```

Comment:
I have moved this block of code to inside the synchronized block. Is that what you mean? Is so, do you mind briefly explaining how moving this into the synchronized block will avoid running into a ConcurrentModificationException?

Reply:
When accessing it without mutual-exclusion, it's possible that a call to `on` or another thing that mutates the collection 
 could happen while the collection is being iterated-over. (for context: for(x in y) and y.forEach both use an `Iterator`).  Enforcing mutual-exclusion means that this will become impossible, since accessing _or_ mutating the collection can only happen on one thread at a time, never in parallel.